# 1. 예외(에러, 오류)
1. 프로그램이 동작하다 어떠한 원인에 의해서 프로그램이강제 종료되는 경우가 있는데, 이런 경우를 초래하는 원인을 에러/오류/예외라고 부른다. 3가지 오류가 있다.
```java 
 try {

 } catch(Exception e) {
        Syso(e.getMessage());
 }
 ```
2. 1. 컴파일에러: 컴파일 중 발생한 에러
2. 2. 런타임에러: 프로그램 실행 중 발생한 에러
2. 3. 논리적에러: 실행은 되는데 결과가 생각대로 동작하지 않는 것

# 2. 예외처리
1. 위처럼 발생한 에러들에 의해 프로그램이 강제종료되는 것을 막거나 에러를 로그로 남겨서 에러가 발생하지 않도록 소스코드를 수정하는 행위 //2가지 방법이 있음
2. 1. Checked Exception: 컴파일러가 예외 처리하는 강제시키는 예외.(무조건 예외처리 코드를 작성해야 하는 예외) IOExeption, ClassNotFoundException 등등이 있으며 throws, try~catch 블록으로 예외처리
2. 2. Unchecked Exceptio: 컴파일러가 예외 처리를 강제하지 않는 예외(예외처리 코드를 작성하지 않아도 됨). NullPointerException, ArrayImdexOutOfBoundException... 등등

# 3. try~catch(~finally) 블록
1. try{} 블록 안의 소스코드를 실행하다가 예외가 발생하면 catch{} 블록으로 이동해서 예외처리 코드를 실행하는 코드. 
2. finally{} 블록은 try에서 예외가 발생하던 발생하지 않던 무조건 실행되는 블록
3. Exception의 getMessage(), printStackTrace()
 - getmessage(): 발생한 예외 클래스의 객체에 저장된 메시지를 전달.
 - printStackTrace(): 예외발생 시 호출된 메소드 정보와 클래스 정보 그리고 예외 메세지까지 모두 출력. 그러나 보안상의 문제로 취약점에 무조건 걸려서 ptintln()이나 log 라이브러리로 대체해서 사용한다. 
4. 예외가 발생했을 때: try{} 블록의 코드 실행 -> 코드 실행 중 예외 발생 시 catch 블록으로 이동
 -> catch{}블록에서 예외처리 -> finally{} 블록 실행
5. 예외가 발생하지 않을 때: try{} 블록의 코드 실행 -> finally{}블록 실행
6. finally{}블록의 작성여부는 선택사항이지만 try{}~catch{}는 무조건 작성 

 # 4. 예외 발생시키기
 1. 예외클래스로 객체를 하나 만들어서, 예외발생
```java
  예외클래스 변수명 = new 예외클래스(///띄워줄 메세지//);
    throw 변수명
```
# 5. 메소등서의 예외 선언 (throws)
1. 메소드를 호출한 곳에서 예외를 전달. 
2. 
```java
리턴타입 메소드명(매개변수) throws 예외클래스 {

}
```

# 6. 예외처리의 중요성
1. 예외처리는 프로그램의 안정성을 보장해줄 수 있는 코드이므로 습관적으로 작성하도록 연습해야 한다. 예외처리 코드를 작성하여 예외가 발생하여도 프로그램이 안정적으로 돌아갈 수 있도록 한다. 
2. try{}~catch{}~finally{} 로 분리되어 있기 대문에 코드 가독성을 높인다. 블록별로 나눠져 있기 때문에 코드의 흐름을 파악하기가 쉽다.  